pgmpy/models/DynamicBayesianNetwork.py

```diff
@@ -533,6 +534,215 @@ def add_cpds(self, *cpds):
+        """
+        Compute the log-likelihood of the given data under the current
+        Dynamic Bayesian Network model.
+
+        The log-likelihood measures how well the model explains the observed data.
+        It is computed as the sum of log-probabilities of all variables at each
+        time slice given their parents (evidence variables). Lower values indicate
+        less probable observations under the current model parameters.
+
+        This method works with discrete Dynamic Bayesian Networks and supports
+        data across multiple time slices. The computation considers:
+        - Variables in the first time slice (t=0) use their marginal probabilities
+        - Variables in subsequent time slices (t>0) use conditional probabilities
+          given their parents from the same and previous time slices
+
+        Parameters
+        ----------
+        data: pd.DataFrame instance
+            The dataset against which to score the model.
+            Each column should represent a variable in the form of (node_name, time_slice).
+
+        show_progress : bool, default=False
+            Whether to display a progress bar during computation.
+
+        Returns
+        -------
+        float
+            The total log-likelihood score for the entire dataset. This is the
+            sum of log-likelihood scores for all individual data points.
+            More negative values indicate worse fit to the data.
+
+        Raises
+        ------
+        ValueError
+            If data is not a pandas DataFrame, contains missing values,
+            has incorrect column structure, or contains values not present
+            in the model's CPD state names.
+
+        Examples
+        --------
+        Examples
+        --------
+        >>> import numpy as np
+        >>> import pandas as pd
+        >>> from pgmpy.models import DynamicBayesianNetwork as DBN
+        >>>
+        >>> # Create a simple DBN with temporal dependencies
+        >>> model = DBN(
+        ...     [(("A", 0), ("B", 0)), (("A", 0), ("A", 1)), (("B", 0), ("B", 1))]
+        ... )
+        >>>
+        >>> # Generate sample data with 4 variables across 3 time slices
+        >>> data = np.random.randint(low=0, high=2, size=(100, 6))
+        >>> colnames = [("A", 0), ("B", 0), ("A", 1), ("B", 1), ("A", 2), ("B", 2)]
+        >>> df = pd.DataFrame(data, columns=colnames)
+        >>>
+        >>> # Fit model and compute log-likelihood
+        >>> model.fit(df)
+        >>> ll_score = model.log_likelihood(df)
+        >>> print(f"Log-likelihood: {ll_score:.2f}")
+        Log-likelihood: -138.62
+        """
+
+        # Input validation
+        if not isinstance(data, pd.DataFrame):
+            raise ValueError(
+                f"Data must be a pandas.DataFrame instance. Got: {type(data)}"
+            )
+
+        if min(data.columns, key=lambda x: x[1])[1] != 0:
+            raise ValueError("Data column names must start from time slice 0.")
+
```

Comment:
What happens if `data.columns` is empty? The `min()` function will raise a `ValueError`. Should we add a check for empty DataFrames before this validation? 

also what if columns are not tuples at all? This will cause a `TypeError` when accessing `x[1]`. Should tuple structure validation happen prior this check?

Reply:
I propose to check if `data` is empty or has empty columns (all) using `data.dropna().empty` property. This way, we ensure `data` has at least one non-empty column so that `min()` won't fail. Then, check for each `data.columns` if the column name is a tuple and has a length of 2, and whether the second item of the tuple is a non-negative integer.

```suggestion
        if data.dropna().empty:
            raise ValueError("DataFrame is empty.")

        for col in data.columns:
            if not isinstance(col, tuple) or len(col) != 2:
                raise ValueError(
                    "Data column names must be tuples of the form (node_name, time_slice). "
                    f"Got column type {type(col)} of length {len(col) if isinstance(col, tuple) else 'N/A'}"
                )
            if not isinstance(col[1], int) or col[1] < 0:
                raise ValueError(
                    "Time slice in data column names must be a non-negative integer."
                )
        
        if min(data.columns, key=lambda x: x[1])[1] != 0:
            raise ValueError("Data column names must start from time slice 0.")
```